File: Online_Judge_Problems/programmers/programmers_389478_택배_상자_꺼내기/pro_389478.py
"""
_-----------------------------------------------------------
Sub    : [Programmers] 택배 상자 꺼내기
Link   : https://school.programmers.co.kr/learn/courses/30/lessons/389478
Level  : 1
Tag    : Python, Math
------------------------------------------------------------
Approach
- 타겟 상자와 마지막 상자의 (row, col) 좌표를 계산한다.
- 타겟 상자 위에 반드시 존재하는 상자의 개수를 계산한다.
- 마지막 층의 같은 열에 상자가 실제 존재하는지 확인한다.
------------------------------------------------------------
"""

import os
import sys

file_path = os.path.join(os.path.dirname(__file__), "input_test.txt")

if os.path.exists(file_path):
    sys.stdin = open(file_path, "r", encoding="utf-8")


# 📥 Input
def get_input_data() -> tuple[int, int, int]:
    n, w, num = map(int, input().split())
    return n, w, num


# ⚙️ Logic
def get_coords(num: int, w: int) -> tuple[int, int]:
    """
    상자 번호를 지그재그 적재 기준의 (row, col) 좌표로 변환한다.

    Args:
        num: 상자 번호 (1-indexed)
        w: 한 줄에 배치되는 상자 개수

    Returns:
        상자의 (row, col) 좌표 (0-indexed)
    """

    row = (num - 1) // w
    remainder = (num - 1) % w

    if row % 2 == 0:
        col = remainder
    else:
        col = w - 1 - remainder

    return row, col


def solution(n: int, w: int, num: int) -> int:
    """
    타겟 상자를 꺼내기 위해 제거해야 하는 상자의 개수를 계산한다.

    Args:
        n: 전체 상자 개수
        w: 한 줄에 배치되는 상자 개수
        num: 꺼낼 상자 번호

    Returns:
        제거해야 하는 상자의 개수
    """

    row, col = get_coords(num, w)
    last_row, _ = get_coords(n, w)

    # 꼭대기 바로 전 층까지는 반드시 상자가 존재한다.
    answer = last_row - row

    if last_row % 2 == 0:
        top_box_num = last_row * w + col + 1
    else:
        top_box_num = last_row * w + (w - 1 - col) + 1

    # 가장 윗층의 같은 열에도 상자가 존재하는 경우
    if top_box_num <= n:
        answer += 1

    return answer


# 🚀 Run Program
if __name__ == "__main__":
    n, w, num = get_input_data()
    print(solution(n, w, num))

File: Online_Judge_Problems/programmers/programmers_389478_택배_상자_꺼내기/pro_389478.ts
/** -----------------------------------------------------------
 * Sub    : [Programmers] 택배 상자 꺼내기
 * Link   : https://school.programmers.co.kr/learn/courses/30/lessons/389478
 * Level  : 1
 * Tag    : TS, Math
 * ------------------------------------------------------------
 * Approach
 * - 상자 번호를 (row, col) 좌표로 변환한다.
 * - 목표 상자와 마지막 상자의 열을 비교하여 위에 존재하는 상자의 개수를 계산한다.
 * ------------------------------------------------------------
 */

declare var require: any;
const fs: any = require("fs");

const filePath: string = fs.existsSync("./input_test.txt")
  ? "./input_test.txt"
  : "/dev/stdin";

const input: string[] = fs.readFileSync(filePath, "utf-8").trim().split(/\s+/);

/* 📥 Get Input Data */
const getInputData = () => {
  const [n, w, num] = input.map(Number);

  return [n, w, num];
};

/* ⚙️ Logic */
/**
 * 상자 번호를 지그재그 적재 기준의 (row, col) 좌표로 변환한다.
 *
 * @param num 타겟 상자 번호 (1-indexed)
 * @param w 한 줄에 배치되는 상자 개수
 * @returns [row, col] (0-indexed)
 */
const getCoords = (num: number, w: number): [number, number] => {
  const row = Math.floor((num - 1) / w);
  const remainder = (num - 1) % w;

  let col: number;

  if (row % 2 === 0) {
    col = remainder;
  } else {
    col = w - 1 - remainder;
  }

  return [row, col];
};

/**
 * 타겟 상자를 꺼내기 위해 제거해야 하는 상자의 개수를 계산한다.
 *
 * @param n 전체 상자 개수
 * @param w 한 줄에 배치되는 상자 개수
 * @param num 꺼낼 상자 번호
 * @returns 제거해야 하는 상자의 개수
 */
const solution = (n: number, w: number, num: number): number => {
  const [row, col] = getCoords(num, w);
  const [lastRow, lastCol] = getCoords(n, w);

  // 꼭대기 바로 전 층까지 존재하는 상자의 개수
  let answer = lastRow - row;

  let topBoxNum: number;

  if (lastRow % 2 === 0) {
    topBoxNum = lastRow * w + col + 1;
  } else {
    topBoxNum = lastRow * w + (w - 1 - col) + 1;
  }

  // 가장 윗층의 같은 열에도 상자가 존재하는 경우
  if (topBoxNum <= n) {
    answer++;
  }

  return answer;
};

/* 🚀 Run Program */
(() => {
  const [n, w, num] = getInputData();
  console.log(solution(n, w, num));
})();
